pkg/image/origins.go

```diff
@@ -0,0 +1,306 @@
+		"ui-plugin-operator":                                      "https://github.com/rancher/ui-plugin-operator",
+		"weave-kube":                                              "https://github.com/weaveworks-experiments/weave-kube",
+		"weave-npc":                                               "https://github.com/weaveworks/weave",
+		"webhook-receiver":                                        "https://github.com/rancher/webhook-receiver",
+		"windows_exporter-package":                                "https://github.com/rancher/windows_exporter-package",
+		"wins":                                                    "https://github.com/rancher/wins",
+		"wmi_exporter-package":                                    "https://github.com/rancher/wmi_exporter-package",
+	}
+}
+
+func GenerateImageOrigins(linuxImagesFromArgs, targetImages, targetWindowsImages []string) error {
+	logrus.Infof("building rancher-image-origins.txt")
+
+	allImages := GetAllUniqueImages(linuxImagesFromArgs, targetImages, targetWindowsImages)
+
+	// if any image does not have a source
+	// declared, or if a net-new image has been
+	// added via KDM or charts and is not
+	// included in the OriginMap
+	// then we cannot continue as the
+	// artifact would be incomplete.
+	if UnknownImages := GatherUnknownImages(allImages); len(UnknownImages) > 0 {
+		return fmt.Errorf("[ERROR] not all images have a source code origin defined. Please provide origin URL's within rancher/pkg/image/origins.go for the following images: \n%s", UnknownImages)
+	}
+
+	// format file
+	fileContents := ""
+	for k, v := range OriginMap {
+		fileContents = fileContents + k + " " + v + "\n"
+	}
+
+	originsFile, err := os.Create("rancher-image-origins.txt")
+	if err != nil {
+		return err
+	}
+
+	originsFile.Chmod(0755)
+	originsFile.WriteString(fileContents)
+	return originsFile.Close()
+}
+
+func GetAllUniqueImages(images ...[]string) []string {
+	var allImages []string
+	for _, imagesSlice := range images {
+		allImages = append(allImages, UniqueTargetImages(imagesSlice)...)
+	}
+	return allImages
+}
+
+func GatherUnknownImages(allImages []string) string {
+	// compare all known images with images gathered
+	// from KDM / Charts to detect images missing from OriginsMap.
+	var unknownImages string
+	for _, e := range allImages {
+		url, ok := OriginMap[e]
+		if !ok || url == "" || url == "unknown" {
+			unknownImages = unknownImages + e + "\n"
+		}
+	}
+	return unknownImages
+}
+
+// repoFromImage strips away the repository and version
+// of a given image.
+func repoFromImage(image string) string {
+	split := strings.Split(image, "/")
+	if len(split) != 2 {
+		return ""
+	}
+
+	split = strings.Split(split[1], ":")
+	if len(split) != 2 {
+		return ""
+	}
+
+	return split[0]
+}
+
+// UniqueTargetImages finds unique images in a list of targetImages,
+// ignoring the image version
+func UniqueTargetImages(targetImages []string) []string {
+	seenImages := make(map[string]bool)
```

Comment:
I think you can replace this with `map[string]struct{}`, which is idiomatic for set containers.

Reply:
I've changed this in two placed, both `UniqueTargetImages` and `CheckForImagesNoLongerBeingUsed`.

Reply:
I've update this for both places where we identify unique images